org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/MatchLocator.java

```diff
@@ -1002,6 +1002,11 @@ public MethodBinding getMethodBinding(MethodPattern methodPattern) {
     				if (unit != null) {
     					AbstractMethodDeclaration amd = new ASTNodeFinder(unit).findMethod((IMethod) methodPattern.focus);
     					if (amd != null && amd.binding != null && amd.binding.isValidBinding()) {
+    						// Check if we "forgot" to resolve parameter types, see: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/821
```

Comment:
I just tried the same workaround for BinaryMethod which is trying to resolve the method by using the following expression as a test `getMethodBinding(methodPattern, ((ProblemReferenceBinding) this.unitScopeTypeBinding).closestReferenceMatch())` and following is the output.

<img width="791" alt="image" src="https://user-images.githubusercontent.com/7471994/229581805-b882550c-8ac1-479e-9421-e84f5ddcf061.png">

Reply:
> I just tried the same workaround for BinaryMethod which is trying to resolve the method by using the following expression as a test `getMethodBinding(methodPattern, ((ProblemReferenceBinding) this.unitScopeTypeBinding).closestReferenceMatch())` and following is the output.

Sorry I don't understand this comment. Do you mean you have an example where the bug from #821 is seen also in a library binary file?

If so, what change are you applying and is the result satisfactory or bad?

Reply:
@trancexpress sorry for not being clear. I checked you changes, if I understand correctly, you try to resolve the method binding with parameters using the fix. 

What i tried is the fix that is there in the if else block, which tries the resolve the method of hidden type because its nested private. 

So in my previous comment the expression i tried out. So if you add a debug point in the if condition which checks for SourceMethod and run that expression, you will get the output in screenshot. 

So does this output is different than what you expect from your fix ? Because if not we can both merge the binary and source method resolutions login in this method right ?

Reply:
As long as the added test and the original problem in #821 are fixed, I don't have a preference. It has been a while so I don't remember the exact specifics (and so I cannot tell from the screenshot whether the bug is fixed). Your suggestion sounds like it will simplify code, so we can try it out. If nothing fails and the problems are fixed, sure, we should do this.

Since you've already adjusted the code on your end, can you amend the commit?

Reply:
@trancexpress not i just did a debug evaluation without all the if guards to see if that provides the method binding with parameters resolved

Reply:
In what context did you try to evaluate `getMethodBinding(methodPattern, ((ProblemReferenceBinding) this.unitScopeTypeBinding).closestReferenceMatch())`?

There is no `ProblemReferenceBinding` in the bug case. So when I evaluate this expression I get:

```
ClassCastException: Cannot cast org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding (id=303) to org.eclipse.jdt.internal.compiler.lookup.ProblemReferenceBinding		
```

Reply:
I did while in debug mode with the tests in this PR. May be you can check adding a conditional breakpoint and see if you every come to this method which satisfies that. Otherwise I can check a bit more tomorrow.

Reply:
OK, I was setting a breakpoint too deep in the source case. When setting a breakpoint at the start of the source case, I see the `ProblemReferenceBinding`.

Can you check the current changes? I've pushed an extra commit so its easier to see the changes.

Unfortunately this is the best I can do. Without the remaining code in the source case there are at least fails in `JavaSearchBugsTests2`. Probably more in other test cases.

Reply:
Though I wonder if we should still make sure `amd.binding` has completed type bindings... I'm not sure if there will be cases where `amd.binding` is returned and its incomplete (due to having no type bindings for its parameters).